burr/visibility/tracing.py

```diff
@@ -321,3 +327,142 @@ def __call__(
             app_id=self.app_id,
             partition_key=self.partition_key,
         )
+
+
+FNType = TypeVar("FNType", bound=Callable)
+
+# This is specifically meant to set anything that is "out of band" -- I.E.
+# Through a decorator that does not have access to the parameter
+# This is an internal API that we are liable to change over time
+# This is set by the application on every `step` call
+
+tracer_factory_context_var: ContextVar[Optional[TracerFactory]] = ContextVar(
+    "tracer_context",
+    default=None,
+)
+
+
+class trace:
+    def __init__(
+        self,
+        capture_inputs: bool = True,
+        capture_outputs: bool = True,
+        input_filterlist: Optional[List[str]] = None,
+        span_name: Optional[str] = None,
+        _context_var: ContextVar[Optional[TracerFactory]] = tracer_factory_context_var,
+    ):
+        """trace() surrounds a burr function and uses the tracer to create a span and log
+        attributes. This also (by default) logs the inputs/outputs of a function as attributes.
+        Be careful not to include sensitive data in the inputs/outputs, but if you do, you have the
+        input_filterlist to exclude it.
+
+        This works with sync/async
+
+        Take the following code:
+
+        .. code-block:: python
+
+            from burr.visibility import trace
+
+            @trace()
+            def call_llm(prompt):
+                return _query(...)
+
+            @trace()
+            def generate_text(prompt: str) -> str:
+                result = call_openai(prompt)
+                return f"<p>{result}</p>"
+
+            @action(reads=["prompt"], writes=["response"])
+            def prompt_action(state: State) -> State:
+                response = generate_text(state["prompt"])
+                return state.update(response=response)
+
+        Every time `prompt_action` is called, it will generate a trace that looks like the following:
+        If it is called outside the context of a Burr action, it will be effectively a no-op.
```

Comment:
```suggestion
        Every time `prompt_action` is called, it will generate a trace that looks like the following:
         
             prompt_action
                  generate_text
                        call_llm
        
        If it is called outside the context of a Burr action, it will be effectively a no-op.
```